File: modules/yandex-map/android/src/main/java/expo/modules/yandexmap/view/CircleView.kt
import android.annotation.SuppressLint
import android.content.Context
import com.yandex.mapkit.geometry.Circle
import com.yandex.mapkit.map.CircleMapObject
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import expo.modules.yandexmap.model.CircleConfig
import expo.modules.yandexmap.model.Coordinate
import expo.modules.yandexmap.utils.ColorUtils
import expo.modules.yandexmap.view.YandexMapView.Companion.mapObjects

@SuppressLint("ViewConstructor")
class CircleView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
    var circleMapObject: CircleMapObject? = null

    private var circleConfig: CircleConfig = CircleConfig()

    fun setCenterValue(center: Coordinate) {
        circleConfig.center = center.toPoint()
    }

    fun setRadiusValue(radius: Float) {
        circleConfig.radius = radius
    }

    fun setStrokeWidthValue(width: Float) {
        circleConfig.strokeWidth = width
    }

    fun setStrokeColorValue(color: String) {
        circleConfig.strokeColor = ColorUtils.parseColor(color)
    }

    fun setFillColorValue(color: String) {
        circleConfig.fillColor = ColorUtils.parseColor(color)
    }

    fun updateCircle() {
        val circle = Circle(
            circleConfig.center!!,
            circleConfig.radius!!
        )

        circleMapObject = mapObjects?.addCircle(circle)?.apply {
            if (circleConfig.strokeWidth != null) {
                strokeWidth = circleConfig.strokeWidth!!
            }

            if (circleConfig.strokeColor != null) {
                strokeColor = circleConfig.strokeColor!!
            }

            if (circleConfig.fillColor != null) {
                fillColor = circleConfig.fillColor!!
            }
        }

    }
}

File: modules/yandex-map/ios/View/CircleView.swift
import ExpoModulesCore
import YandexMapsMobile

class CircleView: ExpoView {
    weak var delegate: MapObjectsDelegate?
    
    var circleMapObject: YMKCircleMapObject? = nil
    
    private var circleConfig: CircleConfig = CircleConfig()
    
    func setCenterValue(_ center: Coordinate) {
        circleConfig.center = center.toPoint()
    }
    
    func setRadiusValue(_ radius: Float) {
        circleConfig.radius = radius
    }
    
    func setStrokeWidthValue(_ strokeWidth: Float) {
        circleConfig.strokeWidth = strokeWidth
    }
    
    func setStrokeColorValue(_ color: String) {
        circleConfig.strokeColor = color
    }
    
    func setFillColorValue(_ color: String) {
        circleConfig.fillColor = color
    }
    
    func updateCircle() {
        guard let mapObjects = delegate?.mapObjects else { return }
        
        let circle = YMKCircle(
            center: circleConfig.center,
            radius: circleConfig.radius
        )
        
        circleMapObject = mapObjects.addCircle(with: circle)
        
        if circleConfig.fillColor != nil {
            circleMapObject!.fillColor = UIColor.fromString(circleConfig.fillColor!)!
        }
        
        if circleConfig.strokeColor != nil {
            circleMapObject!.strokeColor = UIColor.fromString(circleConfig.strokeColor!)!
        }
        
        if circleConfig.strokeWidth != nil {
            circleMapObject!.strokeWidth = circleConfig.strokeWidth!
        }
    }
}
